### ISSUE TITLE
Want to help? Just try out `cargo crev` and give feedback.

### ISSUE THREAD

**Issue by dpc** [dpc/crev#37]
`cargo-crev` is kind of working already. In a sense it's even quite feature complete (alpha quality though)

See https://github.com/dpc/crev/tree/master/cargo-crev for instructions.

**Comment by tylerlaberge** [comment]
Just went through building this project and trying this out and here are some things I ran into (for the most part it seems to be working and looks great)

**Build notes**:

Had some issues building this because of some dependencies I didn't have installed, specifically I did not have OpenSSL installed or a C compiler needed for argonautica. These were resolved with

```
sudo apt-get update
sudo apt-get upgrage

# openssl deps
sudo apt-get install openssl libssl-dev

# argonautica deps
sudo apt-get install clang llvm-dev libclang-dev
```

After getting those dependencies installed the project compiled successfully. Perhaps these dependencies should be called out in the build instructions?

**cargo-crev** notes:

1. _cargo crev verify_ should color code trusted vs not trusted dependencies, currently its a little hard to parse since its all just one color. Green for trusted, Red for not trusted would be cool (maybe yellow for low/medium trusted?)

2. It's a little hard to see what has been reviewed with _cargo crev review_ but not commited/pushed yet. You can do a _cargo crev db git status_ to see files that haven't been pushed, but the filenames don't give any indication of the package the review was for. It would be cool if you could do something like _cargo crev status_ and get a list of dependency names that you have reviewed but not pushed yet.

3. _cargo crev db git push_ does not work for me.

```
$ cargo crev db git push
fatal: The current branch master has no upstream branch.
To push the current branch and set the remote as upstream, use

git push --set-upstream origin master
```

if I try the set-upstream command I get permission denied.

4. Finally, I wonder if the git commands should be kind of abstracted in some way? 

So instead of 

_cargo crev db git add/cargo crev db git commit/cargo crev db git push_

it could instead just be something like

_cargo crev db save (adds/commits)_
_cargo crev db publish (pushes to remote)_

Reason being is I don't exactly see the use case of having all the git commands available, seems like users really would only need the ability to save and publish to me.

Overall, really cool project and I'm excited to see where this goes (and the code looks well written too which is great), hope this feedback helps, when I have time I will try to make some contributions :)

**Comment by dpc** [comment]
I've just improved `cargo crev verify` output . Also created #39 for colors and other improvements.

> It's a little hard to see what has been reviewed

`cargo crev db git diff HEAD` should show you this. Basically all git commands just work, so it should be fairly flexible. If this is not enough, we can definitely add something. It's just the matter of figuring out a consistent usage model.

> ` cargo crev db git push` does not work for me.

#36 - let me know in case it doesn't help (open a new issue with the actual error)

> Finally, I wonder if the git commands should be kind of abstracted in some way?

I'm not sure myself. I am proficient git user, so for me issuing raw git commands is easier, I have my own shortcuts, additional commands etc.

But I guess not every user has to feel this way, so some helpers for most common workflows, could be useful.

**Comment by dpc** [comment]
@tylerlaberge I was rushing to go to sleep yesterday, but I really appreciate the feedback. Thank you!

**Comment by kornelski** [comment]
In the screencast the "and now you review code" part is handwaved. To me, this is the hard part. I know I can't trust package's source code as shown on github, and need to review the actual crate file from crates.io, but getting it is cumbersome. Then I have to ensure I reviewed every file and haven't overlooked anything, and that the crate I dug up is actually the same crate that I'm running trust commands for.

Could you have an interactive mode, similar to `git add -p`, where you display each file in the terminal, and ask "Is this good? (y/n)" and automatically create review data based on that?

**Comment by dpc** [comment]
@kornelski Yes. Of course the review part is the time consuming and difficult one. Part of the reason the `thoroughness` field exists. You can just glance through files, and not even review all, and just mark the `thoroughness` as `none` or `low`. It's still far better than if noone was reviewing them at all. Eventually, when there's enough reviewers and reviews floating around, there will be a time to start getting more methodological about it. I do have some plans for per-file reviews (in form of *Code Review Proofs*) but it's not complete atm. We could also add some helpful tooling to at least help mark files which were reviewed for the convenience of the reviewer.

**Comment by mglax** [comment]
Works fine on Mac OS X. One thing that stroke me first is that the "rating" should be split, the security aspect and the functional aspects are, to me, not really related. I am reviewing a deprecated crate since it appears first on crates.io (the name is an exact match): there are no security flaws that I can see (positive review), but since it's deprecated, I'd put a negative note for the functional part.

**Comment by dpc** [comment]
@mglax I don't really see how "functional but insecure" or "secure but non functional" states are any useful for the downstream user of the review. IMO, in both cases you just want to give `rating: negative` or `rating: dangerous` and describe in the `comment` what's wrong. For the mass-consumers, all of this will go through WoT graph and will have to spit out "OK to use" or "not OK to use" decision anyway. 

Once the downstream user will get a dependency flagged for any reason, they are supposed to read the offending comments, and judge accordingly what to do, potentially overwriting the `status` with their own review. Eg. `rating: neutral`, `comment: Yeah, there are problems, but for my particular purpose it's OK`.

I'm open for discussion about it though.

**Comment by daxpedda** [comment]
If I only want to use `cargo crev verify deps`, why do I have to generate an ID first, which also implies me having to create a `crev-proofs` repo?
Shouldn't it be enough to just `fetch` proofs from your trusted source and then run verify?

I had a great experience otherwise. Wish this would be part of the official rust ecosystem. We really needed this desperately already yesterday.

**Comment by dpc** [comment]
@daxpedda You're right. I've created #107 to fix it. Thanks!

**Comment by dpc** [comment]
@daxpedda Fixed. I'd appreciate a lot if you could test the current git version. Thanks!

**Comment by daxpedda** [comment]
Is there a way to log out? Is there a way to find out where currently fetched proofs are located?

I can create seperate issues for those if you want.

**Comment by dpc** [comment]
@daxpedda "Log out"? What exactly do you mean? 

Fetched proofs are stored in subdirectories for `~/.cache/crev/remotes/<urlid>`. It is being displayed when `crev fetch <smth>` is used.

Your local proof repo is in `~/.config/crev/proofs/<urlid>`.

In both cases the `<urlid>` is a base64 (url-safe) encoded blake hash of the url.

**Comment by daxpedda** [comment]
> @daxpedda "Log out"? What exactly do you mean?

After deleting the cache folder, if I `cargo crev query id current` it will still tell me who I am and my ID. Is there a way to reset it? Like `cargo crev switch none` or something like that?

> Fetched proofs are stored in subdirectories for `~/.cache/crev/remotes/<urlid>`.

On Windows it saves it in `X:\Users\<Username>\AppData\Local\Dawid Ci,281,,380,arkiewicz\crev\remotes`, is that intended? Will that be changed? Should it be documented somewhere?

**Comment by dpc** [comment]
@daxpedda I've created #124 to address the "not-initialied". Initially `crev` was meant to require ID in all cases, and I still haven't changed it for all the commands.

There are two concepts here: one is cache of where `cargo crev fetch ...` stuff goes. The other one is local copy of your git proof repo (where your own proofs go). Your own repo repo might or might not happen to be fetched into a cache too. We're using `app_dirs` to create these dirs, I belive, and if you think on Windows these should be some other places, placase create a separate issue and let me know what do you think, or even better - submit a PR. I'm not a Windows user so Windows support will require some good people to step in and help out. :)

Right now there's no command to "unswitch" from an ID. I'm not sure if this is really needed. If you want to try things without any ID, I guess creating a new, fresh ID with invalid url would effectively do that. You can also delete the whole local crev diff (and possibly cache is well, if you really want).

I guess we could have a command for cleaning up the cache ...

Can you explain why exactly do you need this things? I knew, maybe there are some better ways to support such use case.

**Comment by dpc** [comment]
> So I was thinking to use crev in my CI.

I totally want it, but just didn't get to implement it yet.  Created #131

**Comment by dpc** [comment]
@arbitrix https://github.com/dpc/crev/tree/master/cargo-crev#windows

If you can think of more specific instructions, I would be happy to land a PR with them. :) I myself, don't have any Windows box, so I rely on other devs/users for Windows support.
